src/ModelObjects/component.rs

```diff
@@ -255,6 +255,218 @@ impl Component {
         self.output_edges = Some(o_edges);
         self.input_edges = Some(i_edges);
     }
+
+    /// Function for reducing the clocks found on the component.
+    /// Unused clocks and "duplicate" clocks (clocks that are never reset)
+    /// and then remove them.
+    pub fn reduce_clocks(&mut self, redundant_clocks: &Vec<RedundantClock>) {
+        for clock in redundant_clocks {
+            match &clock.reason {
+                ClockReductionReason::Duplicate(global) => self.replace_clock(&clock, global),
+                ClockReductionReason::Unused => self.remove_clock(&clock.updates),
+            }
+
+            let clock_val = *self.declarations.clocks.get(clock.clock.as_str()).unwrap();
+            self.declarations
+                .clocks
+                .values_mut()
+                .filter(|val| **val > clock_val)
+                .for_each(|val| *val -= 1);
+            self.declarations.clocks.remove(clock.clock.as_str());
+        }
+    }
+
+    ///Used to find redundant clocks - checks for unused and duplicates clocks.
+    ///
+    /// Returns [`Vec<RedundantClock>`] with all found redundant clock.
+    /// If no redundant clocks found the vector will be empty
+    pub(crate) fn find_redundant_clocks(&self) -> Vec<RedundantClock> {
+        let clocks: HashSet<String> = self.declarations.get_clocks().keys().cloned().collect();
+        let mut out: Vec<RedundantClock> = vec![];
+        let mut seen_clocks: HashMap<String, Box<[Vec<usize>; 2]>> = HashMap::new();
+        for (index, expr, which) in self
```

Comment:
The forloop looks pretty but maybe little hard to read whats happening.  Still unsure out errorhandling all the unwarps

Reply:
Should I make a few comments?

Reply:
done